File: clock/theme_loader.py
from __future__ import annotations

import json
import os
import threading
from dataclasses import dataclass, field
from typing import Optional

try:
    from watchdog.observers import Observer
    from watchdog.events import FileSystemEventHandler
    _WATCHDOG_AVAILABLE = True
except ImportError:
    _WATCHDOG_AVAILABLE = False


@dataclass
class Theme:
    name: str
    description: str = ""
    background_type: str = "solid"
    background_color: str = "#000820"
    background_top: str = "#0F0019"
    background_bottom: str = "#3C1400"
    background_split: float = 0.5
    background_gradient_direction: str = "sunset"
    colors: dict = field(default_factory=dict)
    stars_enabled: bool = False
    shooting_stars_enabled: bool = False
    cameos: list = field(default_factory=list)
    cloud_density: str = "medium"
    cloud_speed: str = "medium"
    sun_enabled: bool = True
    moon_enabled: bool = False
    precipitation: str = "none"
    condition_overrides: dict = field(default_factory=dict)


def _theme_from_dict(data: dict) -> Theme:
    known = {f.name for f in Theme.__dataclass_fields__.values()}  # type: ignore[attr-defined]
    filtered = {k: v for k, v in data.items() if k in known}
    theme = Theme(**filtered)
    # Backward compatibility: shooting_stars_enabled → cameo entry
    if data.get("shooting_stars_enabled") and not any(
        c.get("name") == "shooting_star" for c in theme.cameos
    ):
        theme.cameos = theme.cameos + [{"name": "shooting_star", "chance_per_minute": 8}]
    return theme


class ThemeLoader:
    def __init__(self, themes_dir: str):
        self.themes_dir = themes_dir
        self.on_themes_changed = None
        self._themes: dict[str, Theme] = {}
        self._observer = None
        self._handler = None

        os.makedirs(themes_dir, exist_ok=True)
        self._themes = self.load_all()
        self._start_watcher()

    def load_all(self) -> dict[str, Theme]:
        themes = {}
        for entry in os.scandir(self.themes_dir):
            if not entry.name.endswith(".json") or not entry.is_file():
                continue
            try:
                with open(entry.path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                if "name" not in data:
                    data["name"] = entry.name[:-5]
                theme = _theme_from_dict(data)
                themes[theme.name] = theme
            except Exception as e:
                print(f"[theme_loader] Warning: skipping {entry.name}: {e}")
        return themes

    def get_theme(self, name: str) -> Optional[Theme]:
        return self._themes.get(name)

    def available_themes(self) -> list[str]:
        return sorted(self._themes.keys())

    def _reload(self):
        self._themes = self.load_all()
        if callable(self.on_themes_changed):
            self.on_themes_changed(dict(self._themes))

    def _start_watcher(self):
        if not _WATCHDOG_AVAILABLE:
            print("[theme_loader] Warning: watchdog not installed — file watching disabled")
            return

        loader = self

        class _Handler(FileSystemEventHandler):
            def __init__(self):
                super().__init__()
                self._timer = None
                self._lock = threading.Lock()

            def on_any_event(self, event):
                if event.is_directory:
                    return
                if not (getattr(event, "src_path", "").endswith(".json") or
                        getattr(event, "dest_path", "").endswith(".json")):
                    return
                with self._lock:
                    if self._timer is not None:
                        self._timer.cancel()
                    self._timer = threading.Timer(0.5, loader._reload)
                    self._timer.daemon = True
                    self._timer.start()

        self._handler = _Handler()
        self._observer = Observer()
        self._observer.schedule(self._handler, self.themes_dir, recursive=False)
        self._observer.daemon = True
        self._observer.start()

    def stop(self):
        if self._handler is not None:
            with self._handler._lock:
                if self._handler._timer is not None:
                    self._handler._timer.cancel()
                    self._handler._timer = None
        if self._observer is not None:
            self._observer.stop()
            self._observer.join()
            self._observer = None
        self._handler = None

File: clock-cpp/src/ThemeLoader.cpp
#include "Theme.h"
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <iostream>

namespace fs = std::filesystem;

// ── JSON → Theme ──────────────────────────────────────────────────────────────

Theme themeFromJson(const nlohmann::json& j) {
    Theme t;

    t.name        = j.value("name", "");
    t.description = j.value("description", "");

    t.backgroundType              = j.value("background_type",               "solid");
    t.backgroundColor             = j.value("background_color",              "#000820");
    t.backgroundTop               = j.value("background_top",                "#0F0019");
    t.backgroundBottom            = j.value("background_bottom",             "#3C1400");
    t.backgroundSplit             = j.value("background_split",              0.5f);
    t.backgroundGradientDirection = j.value("background_gradient_direction", "sunset");

    if (j.contains("colors") && j["colors"].is_object()) {
        for (auto& [k, v] : j["colors"].items()) {
            t.colors[k] = parseColorValue(v);
        }
    }

    t.starsEnabled         = j.value("stars_enabled",          false);
    t.shootingStarsEnabled = j.value("shooting_stars_enabled", false);

    if (j.contains("cameos") && j["cameos"].is_array()) {
        for (const auto& ce : j["cameos"]) {
            CameoEntry entry;
            entry.name            = ce.value("name",              "");
            entry.chancePerMinute = ce.value("chance_per_minute", 0);
            entry.raw             = ce.is_object() ? ce : nlohmann::json::object({{"name", entry.name}});
            t.cameos.push_back(entry);
        }
    }

    // Backward compat: shooting_stars_enabled → cameo entry (mirrors Python).
    if (t.shootingStarsEnabled) {
        bool hasEntry = false;
        for (const auto& c : t.cameos) {
            if (c.name == "shooting_star") { hasEntry = true; break; }
        }
        if (!hasEntry) {
            t.cameos.push_back({"shooting_star", 8, nlohmann::json::object({
                {"name", "shooting_star"},
                {"chance_per_minute", 8},
            })});
        }
    }

    if (t.starsEnabled) {
        bool hasEntry = false;
        for (const auto& c : t.cameos) {
            if (c.name == "stars") { hasEntry = true; break; }
        }
        if (!hasEntry) {
            t.cameos.insert(t.cameos.begin(), {"stars", 0, nlohmann::json::object({
                {"name", "stars"},
            })});
        }
    }

    t.cloudDensity = j.value("cloud_density", "medium");
    t.cloudSpeed   = j.value("cloud_speed",   "medium");
    t.sunEnabled   = j.value("sun_enabled",   true);
    t.moonEnabled  = j.value("moon_enabled",  false);
    t.precipitation = j.value("precipitation", "none");

    if (j.contains("condition_overrides") && j["condition_overrides"].is_object()) {
        for (auto& [k, v] : j["condition_overrides"].items()) {
            t.conditionOverrides[k] = v;
        }
    }

    return t;
}

// ── ThemeLoader ───────────────────────────────────────────────────────────────

ThemeLoader::ThemeLoader(std::string themesDir)
    : _themesDir(std::move(themesDir))
{
    fs::create_directories(_themesDir);
    loadAll();
}

ThemeLoader::~ThemeLoader() = default;

void ThemeLoader::loadAll() {
    _themes.clear();
    std::error_code ec;
    for (const auto& entry : fs::directory_iterator(_themesDir, ec)) {
        if (entry.path().extension() != ".json") continue;
        try {
            std::ifstream f(entry.path());
            nlohmann::json j = nlohmann::json::parse(f);
            if (!j.contains("name")) {
                j["name"] = entry.path().stem().string();
            }
            Theme theme = themeFromJson(j);
            _themes[theme.name] = std::move(theme);
        } catch (const std::exception& e) {
            std::cerr << "[theme_loader] Warning: skipping "
                      << entry.path().filename() << ": " << e.what() << "\n";
        }
    }
    if (ec) {
        std::cerr << "[theme_loader] Warning: " << _themesDir << ": " << ec.message() << "\n";
    }
}

std::optional<Theme> ThemeLoader::getTheme(const std::string& name) const {
    auto it = _themes.find(name);
    if (it == _themes.end()) return std::nullopt;
    return it->second;
}

std::vector<std::string> ThemeLoader::availableThemes() const {
    std::vector<std::string> names;
    names.reserve(_themes.size());
    for (const auto& [k, _] : _themes) names.push_back(k);
    std::sort(names.begin(), names.end());
    return names;
}
